src/graphql/mirror.js

```diff
@@ -684,6 +697,236 @@ export class Mirror {
+      return (args) => {
+        const result = stmt.run(args);
+        // istanbul ignore if: should not be reachable
+        if (result.changes !== 1) {
+          throw new Error(
+            "Invariant violation: bad change count: " +
+              JSON.stringify({stmt, args})
+          );
+        }
+      };
+    };
+
+    // Update all primitive values for an object.
+    const updatePrimitives: ({|
+      +id: Schema.ObjectId,
+      +[primitiveFieldName: Schema.Fieldname]: string,
+    |}) => void = (() => {
+      if (objectType.primitiveFieldNames.length === 0) {
+        return () => {};
+      }
+      const tableName = _primitivesTableName(typename);
+      const updates = objectType.primitiveFieldNames
+        .map((f) => `"${f}" = :${f}`)
+        .join(", ");
+      const stmt = db.prepare(
+        `UPDATE ${tableName} SET ${updates} WHERE id = :id`
+      );
+      return makeUpdateFunction(stmt);
+    })();
+
+    // Update a single link value for an object.
+    const updateLink: ({|
+      +parentId: string,
+      +fieldname: string,
+      +childId: string | null,
+    |}) => void = (() => {
+      const stmt = db.prepare(
+        dedent`\
+          UPDATE links SET child_id = :childId
+          WHERE parent_id = :parentId AND fieldname = :fieldname
+        `
+      );
+      return makeUpdateFunction(stmt);
+    })();
+
+    // Set the `last_update` field of the given object to the `updateId`
+    // provided to the enclosing function.
+    const setLastUpdate: (objectId: Schema.ObjectId) => void = (() => {
+      const stmt = db.prepare(
+        dedent`\
+          UPDATE objects SET last_update = :updateId
+          WHERE id = :objectId
+        `
+      );
+      const update = makeUpdateFunction(stmt);
+      return (objectId) => update({objectId, updateId});
+    })();
+
+    const doesObjectExist = db
+      .prepare("SELECT COUNT(1) FROM objects WHERE id = ?")
+      .pluck();
+    for (const entry of queryResult) {
+      if (!doesObjectExist.get(entry.id)) {
+        throw new Error(
+          "Cannot update data for nonexistent node: " + JSON.stringify(entry.id)
+        );
+      }
+    }
+
+    for (const entry of queryResult) {
+      if (entry.__typename !== typename) {
+        const s = JSON.stringify;
+        throw new Error(
+          "Result set has inconsistent typenames: " +
+            `${s(typename)} vs. ${s(entry.__typename)}`
+        );
+      }
+      setLastUpdate(entry.id);
+      for (const fieldname of objectType.linkFieldNames) {
+        const value: PrimitiveResult | NodeFieldResult = entry[fieldname];
+        const linkData: NodeFieldResult = (value: any);
+        if (linkData === undefined) {
+          const s = JSON.stringify;
+          throw new Error(
+            `Missing node reference ${s(fieldname)} on ${s(entry.id)} ` +
+              `of type ${s(typename)} (got ${(linkData: empty)})`
+          );
+        }
+        let childId = null;
```

Comment:
possibly we can abstract over this pattern? it seems kind of generic and we might like to fold it out of this particular update